make eta configurable

jduerholt opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to make BoTorch better.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to BoTorch here: https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md
-->

## Motivation

I recently looked into the output constraint implementation in botorch and figured out that it behave like our custom `Objective` implementation for handling of constraints, namely by multiplying by sigmoids. Currently, the only difference is that we work often with different `eta` values per constraint. I think this would be a nice feature also for `botorch`. 

This PR is still work in progress, as the `apply_constraints` method is used at a lot of different occasions though-out the codebase, and my question is if one want to keep backwards compatibility. In the current PR, I kept the backwards compatibility and made `eta` of type `Union[float, torch.Tensor]`. If one does this one has to always catch if just a float is provided and transform the float in a tensor of the same length as the list of constraint callables.

Another option would be to set `eta` as optional with default `None` and then just generate a tensor with the old default of 10e-3. 

Which solution would you prefer?

Depending on your suggestion, I would finalize the PR and implement the functionality of different `eta`s per constraint though-out the whole codebase. 

### Have you read the [Contributing Guidelines on pull requests]

Yes.

## Test Plan

Unit tests.

saitcakmak commented on botorch/acquisition/objective.py:
```diff
@@ -472,6 +472,8 @@ def __init__(
         super().__init__(objective=objective)
         self.constraints = constraints
+        if type(eta) != Tensor:
+            eta = [eta for _ in range(len(constraints))]
         self.register_buffer("eta", torch.as_tensor(eta))
```
Nit: You can use `torch.full` to construct the tensor in one step.
```suggestion
            eta = torch.full(len(constraints), eta)
        self.register_buffer("eta", eta)
```

saitcakmak commented on botorch/utils/objective.py:
```diff
@@ -83,9 +83,11 @@ def apply_constraints_nonnegative_soft(
         A `n_samples x b x q (x m')`-dim tensor of feasibility-weighted objectives.
     """
+    if type(eta) != Tensor:
+        eta = torch.as_tensor([eta for _ in range(len(constraints))])
```
```suggestion
        eta = torch.full(len(constraints), eta)
```

saitcakmak commented on botorch/utils/objective.py:
```diff
@@ -118,7 +120,7 @@ def apply_constraints(
     samples: Tensor,
     infeasible_cost: float,
-    eta: float = 1e-3,
+    eta: Union[List[float], float] = 1e-3,
```
Why is it `List[float]` here and `Tensor` everywhere else?

jduerholt replied:
I forgot to update it, I will change it.

jduerholt replied:
Fixed.

saitcakmak commented on botorch/utils/objective.py:
```diff
@@ -83,9 +83,11 @@ def apply_constraints_nonnegative_soft(
     obj = obj.clamp_min(0)  # Enforce non-negativity with constraints
-    for constraint in constraints:
-        constraint_eval = soft_eval_constraint(constraint(samples), eta=eta)
+    for constraint, e in zip(constraints, eta):
+        constraint_eval = soft_eval_constraint(constraint(samples), eta=e)
```
What happens here if `eta` is a single element tensor? I think this would ignore all the constraints after the first one. We should either add a check that they're the same length and error out, or extend the if statement above to catch single element tensors.

jduerholt replied:
I forgot to update it, will change it ;)

jduerholt replied:
I think the best is to just error out.

jduerholt replied:
I fixed it by catching it in `apply_constraints_nonnegative_soft`.

saitcakmak reviewed (changes requested):
Thanks for putting this together. I think being backwards compatible would be preferable, so let's accept both Tensor & float. I left a few comments in line. This should be good to go if you update the docstrings (to explain what the tensor version is) and add units.

jduerholt commented:
Thanks for the quick review, I will go through the codebase and change it accordingly. Best, Johannes

jduerholt commented:
Another thing which I just recognized: I think the doc string for `soft_eval_constraint` is wrong. It is stated: `As eta grows larger, this approximates the Heaviside step function.` The opposite is the case: as eta decreases, it approximates the Heaviside step function, or?

```python

from botorch.utils.objective import soft_eval_constraint, apply_constraints
import matplotlib.pyplot as plt
import numpy as np
import torch

x = np.linspace(-10,10,500)
xt = torch.from_numpy(x).unsqueeze(-1)

fig, ax = plt.subplots()

y = soft_eval_constraint(xt, eta=1e-3 )
ax.plot(xt, y, label="eta = 10e-3")
y = soft_eval_constraint(xt, eta=1e-2 )
ax.plot(xt, y, label="eta = 10e-2")
y = soft_eval_constraint(xt, eta=1e-1 )
ax.plot(xt, y, label="eta = 10e-1")
y = soft_eval_constraint(xt, eta=1 )
ax.plot(xt, y, label="eta = 1")

ax.legend()
plt.show()
```
This gives the following output:

![image](https://user-images.githubusercontent.com/28678090/204520885-6ba28281-9e39-4ba4-8de7-ad9265ebab46.png)


Or do I understand something wrong?

Balandat commented:
You're right, the docstring is wrong! Would you mind patching it as part of this PR?

jduerholt commented:
> You're right, the docstring is wrong! Would you mind patching it as part of this PR?

I just fixed it. Internally we call the inverse of the `eta` parameter `steepness`, which makes it at least for me clearer what it is doing.

jduerholt commented:
I added the possibility for different `eta`s now to the `ConstrainedMCObjective` and into qehvi/qnehevi and the derived methods. 

In addition, I added tests, for qnehvi one test is failing, namely if one uses an eta of 0.1 or 1. Then it shows a strange behavior. For eta=1, it even results in values larger than with the unconstrained one. This should not be. Maybe one of you can have a look, as I do not understand what is going on. The test is failing here: https://github.com/jduerholt/botorch/blob/bb5a56088759302041e7c818f4efe33559906c36/test/acquisition/multi_objective/test_monte_carlo.py#L1301 

Any ideas?

saitcakmak commented on botorch/acquisition/multi_objective/monte_carlo.py:
```diff
@@ -85,6 +85,7 @@ def __init__(
         sampler: Optional[MCSampler] = None,
         objective: Optional[MCMultiOutputObjective] = None,
         constraints: Optional[List[Callable[[Tensor], Tensor]]] = None,
+        eta: Optional[Union[Tensor, float]] = 1e-3,
```
Let's add a docstring for this as well.

jduerholt replied:
Done.

saitcakmak commented on botorch/utils/objective.py:
```diff
@@ -78,14 +78,24 @@ def apply_constraints_nonnegative_soft(
+        eta = torch.full((len(constraints),), eta)
+    if len(eta) != len(constraints):
+        raise ValueError(
+            "Number of provided constraints and numer of provided etas does not match."
```
```suggestion
            "Number of provided constraints and number of provided etas do not match."
```

saitcakmak reviewed (commented):
I'll take a look at the test failure and follow up.

saitcakmak commented on botorch/acquisition/multi_objective/multi_fidelity.py:
```diff
@@ -46,9 +46,9 @@ def __init__(
         sampler: Optional[MCSampler] = None,
         objective: Optional[MCMultiOutputObjective] = None,
         constraints: Optional[List[Callable[[Tensor], Tensor]]] = None,
+        eta: Optional[Union[Tensor, float]] = 10e-3,
```
```suggestion
        eta: Optional[Union[Tensor, float]] = 1e-3,
```

saitcakmak commented on test/acquisition/multi_objective/test_monte_carlo.py:
```diff
@@ -527,6 +527,65 @@ def test_constrained_q_expected_hypervolume_improvement(self):
                 )
                 res = acqf(X)
                 self.assertAlmostEqual(res.item(), 0.5 * 1.5, places=4)
+            # test zero slack multiple constraints
+            for eta in (1e-1, 1e-2):
+                acqf = qExpectedHypervolumeImprovement(
+                    model=mm,
+                    ref_point=ref_point,
+                    partitioning=partitioning,
+                    sampler=sampler,
+                    constraints=[
+                        lambda Z: torch.zeros_like(Z[..., -1]),
+                        lambda Z: torch.zeros_like(Z[..., -1]),
+                    ],
+                    eta=eta,
+                )
+                res = acqf(X)
+                self.assertAlmostEqual(res.item(), 0.5 * 0.5 * 1.5, places=4)
+            # test multiple constraints one eta
+            acqf = qExpectedHypervolumeImprovement(
+                model=mm,
+                ref_point=ref_point,
+                partitioning=partitioning,
+                sampler=sampler,
+                constraints=[
+                    lambda Z: torch.ones_like(Z[..., -1]),
+                    lambda Z: torch.ones_like(Z[..., -1]),
+                ],
+                eta=1.0,
+            )
+            res = acqf(X)
+            self.assertAlmostEqual(
+                res.item(),
+                (
+                    torch.sigmoid(torch.as_tensor(-1.0))
+                    * torch.sigmoid(torch.as_tensor(-1.0))
+                    * 1.5
+                ).item(),
+                places=4,
+            )
```
It'd be good to reduce the repetition in these test as well. Looks like only eta and the expected value changes. Maybe you can run them in a for loop?

jduerholt replied:
Yeah, I introduced loops for this. In principle one could have it even more concise but then one would loose readability. Personally, I really like the `@pytest.mark.parametrize` decorator for these kind of tests, as they make the code very condense but still readable. But I think you are not using it in botorch.

saitcakmak commented on test/acquisition/multi_objective/test_monte_carlo.py:
```diff
@@ -1074,7 +1133,50 @@ def test_constrained_q_noisy_expected_hypervolume_improvement(self):
+                mm._posterior._samples = samples
+                res = acqf(X)
+                self.assertAlmostEqual(res.item(), 0.5 * 0.5 * 1.5, places=4)
+            # test zero slack multiple constraints, multiple etas
+            # set the MockPosterior to use samples over baseline points
+            mm._posterior._samples = baseline_samples
+            sampler = IIDNormalSampler(sample_shape=torch.Size([1]))
+            acqf = qNoisyExpectedHypervolumeImprovement(
+                model=mm,
+                ref_point=ref_point,
+                X_baseline=X_baseline,
+                sampler=sampler,
+                constraints=[
+                    lambda Z: torch.zeros_like(Z[..., -1]),
+                    lambda Z: torch.zeros_like(Z[..., -1]),
+                ],
+                eta=torch.tensor([1.0, 10.0]),
+                cache_root=False,
+            )
+            # set the MockPosterior to use samples over baseline points and new
+            # candidates
+            mm._posterior._samples = samples
+            res = acqf(X)
+            self.assertAlmostEqual(res.item(), 0.5 * 0.5 * 1.5, places=4)
```
looks like this could also go in the for loop.

jduerholt replied:
Done.

saitcakmak commented on test/acquisition/multi_objective/test_monte_carlo.py:
```diff
@@ -1169,6 +1271,38 @@ def test_constrained_q_noisy_expected_hypervolume_improvement(self):
+            # test multiple constraints one eta with
+            # this crashes for large etas, and I do not why
+            # set the MockPosterior to use samples over baseline points
+            mm._posterior._samples = baseline_samples
+            sampler = IIDNormalSampler(sample_shape=torch.Size([1]))
+            acqf = qNoisyExpectedHypervolumeImprovement(
+                model=mm,
+                ref_point=ref_point,
+                X_baseline=X_baseline,
+                sampler=sampler,
+                constraints=[
+                    # lambda Z: torch.zeros_like(Z[..., -1]),
+                    lambda Z: torch.ones_like(Z[..., -1]),
+                ],
+                eta=1,
+                cache_root=False,
+            )
+            samples = torch.cat(
+                [
+                    baseline_samples.unsqueeze(0),
+                    torch.tensor([[[6.5, 4.5]]], **tkwargs),
+                ],
+                dim=1,
+            )
+            mm._posterior._samples = samples
```
Re test failure: It seems like you expect the infeasible acqf value to be the feasible acqf value times the sigmoid value of the constraints there. This does not work there, since the two acqfs compute the hypervolume improvement w.r.t two different partitionings. The feasible acqf has a partitioning that has a non-empty set of baseline points (you can print `acqf.partitioning.Y`) but the infeasible acqf has a partitioning that is empty. This difference happens because the constraints are taken into account while constructing the partitioning in `__init__ -> _set_cell_bounds`. 

If instead of constructing a new acqf in here, you just update the constraints and eta as below, the test will pass.
```suggestion
            acqf.constraints = [
                # lambda Z: torch.zeros_like(Z[..., -1]),
                lambda Z: torch.ones_like(Z[..., -1]),
            ]
            acqf.eta = torch.tensor([1.0])
```

jduerholt replied:
Thank you very much for the explanation! I updated the tests accordingly.

jduerholt commented:
Hi @saitcakmak , thank you very much for your suggestions and especially your explanation regarding the base line partioning for qnehvi! I updated the tests accordingly, just tell me if more is needed.

saitcakmak reviewed (approved):
Thanks for making all the changes and thorough testing. LGTM!

Balandat commented on test/acquisition/multi_objective/test_monte_carlo.py:
```diff
@@ -1169,6 +1234,37 @@ def test_constrained_q_noisy_expected_hypervolume_improvement(self):
+            # this crashes for large etas, and I do not why
+            # set the MockPosterior to use samples over baseline points
+            etas = [torch.tensor([1.0]), torch.tensor([1.0, 10.0])]
+            constraints = [
+                [lambda Z: torch.ones_like(Z[..., -1])],
+                [
+                    lambda Z: torch.ones_like(Z[..., -1]),
+                    lambda Z: torch.ones_like(Z[..., -1]),
+                ],
+            ]
```
Could make this a bit cleaner, but not really necessary
```suggestion
            
            def con_func(Z):
                torch.ones_like(Z[..., -1])
            
            constraints = [[con_func], [con_func, con_func]]
```

jduerholt replied:
Hi @Balandat , if you want I can change it accordingly, is this still possible as the PR is already imported to Phabricator?

Balandat replied:
Yeah we can always reimport (not anymore for this one b/c it already got merged). So no worries for this one as this was really a minor nit.
